### PR TITLE
✨ ClusterClass: Introduce NamingStrategy and allow generating names using go templates

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Introduces the NamingStrategy struct at the ClusterClass level for:
* ControlPlane
* MachineDeployments
* MachinePools

Also introduces a new package which exposes functions and a interface to generate names.

TODO: improve this message here to have more details

TODO:

- [x] improve godocs
  - [x] add examples of which template data struct to expect on the godocs
  - [x] review all new godocs
- [x] unit tests

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #9323

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area clusterclass

### CODE REVIEW

File: api/v1beta1/clusterclass_types.go [hunk 1]
```diff
@@ -637,6 +649,11 @@ type LocalObjectTemplate struct {
 	Ref *corev1.ObjectReference `json:"ref"`
 }
 
+// NamingStrategy defines a naming strategy for templated objects.
+type NamingStrategy struct {
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply api/v1beta1/clusterclass_types.go:653]: Hey very good ask!

I just had a short discussion with @sbueringer about this and I think there are three options:


Option 1 (as currently is on this PR):

```yaml
spec:
  ...
  controlPlane:
    namingStrategy:
      name: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplateName: "{{ .cluster.name }}-{{ .randomSuffix }}"
```

Option 2: similar to what CAPI already does for variables

```yaml
spec:
  ...
  controlPlane:
    namingStrategy:
      name:
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplateName:
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
```


Option 3 (what you proposed): similar to Option 2 but adding a type:

```yaml
spec:
  ...
  controlPlane:
    namingStrategy:
      name:
        type: Template
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplateName:
        type: Template
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
```

I lean towards Option 2. It is also flexible to move to option 3 in future if we introduce other variants..

Note: This also adds a `machineInfrastructureTemplateName`, to show how it could get expanded to other resources. In scope for implementation for now is only `name`.

[the remark this answers is not in the document]
Comment (sbueringer) [reply api/v1beta1/clusterclass_types.go:653]: +1 for Option 2

File: api/v1beta1/clusterclass_types.go [hunk 2]
```diff
@@ -637,6 +649,11 @@ type LocalObjectTemplate struct {
 	Ref *corev1.ObjectReference `json:"ref"`
 }
 
+// NamingStrategy defines a naming strategy for templated objects.
```
Comment (sbueringer) [inline api/v1beta1/clusterclass_types.go:652]: We should probably use different structs so we can evolve it independently (e.g. when we want to add other names that don't exist for CP/MD/MP in the same way)

(but let's to this after we got consensus on the API)

File: internal/controllers/topology/cluster/desired_state.go [hunk 3]
```diff
@@ -696,13 +707,23 @@ func computeMachineDeployment(ctx context.Context, s *scope.Scope, machineDeploy
...
 	}
 
+	nameTemplate := "{{ .cluster.name }}-{{ .machineDeployment.topologyName }}-{{ .random }}"
+	if machineDeploymentClass.NamingStrategy != nil && machineDeploymentClass.NamingStrategy.Name != nil {
+		nameTemplate = *machineDeploymentClass.NamingStrategy.Name
+	}
+
+	name, err := names.MachineDeploymentNameGenerator(nameTemplate, s.Current.Cluster.Name, machineDeploymentTopology.Name).GenerateName()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generator name for machine deployment")
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state.go:717]:
```suggestion
		return nil, errors.Wrap(err, "failed to generate name for machine deployment")
```

Comment (sbueringer) [reply internal/controllers/topology/cluster/desired_state.go:717]: Same below

File: internal/controllers/topology/names/namegenerator.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	randomLength           = 5
+	maxGeneratedNameLength = maxNameLength - randomLength
+)
+
+type simpleNameGenerator struct {
+	base string
+}
+
+func (s *simpleNameGenerator) GenerateName() (string, error) {
+	return names.SimpleNameGenerator.GenerateName(s.base), nil
```
Comment (sbueringer) [inline internal/controllers/topology/names/namegenerator.go:40]: Not sure but maybe we should just inline the GenerateName func of the SimpleNameGenerator

This way this uses the same constants as the code below

File: internal/test/builder/builders.go [hunk 5]
```diff
@@ -485,6 +486,12 @@ func (m *MachineDeploymentClassBuilder) WithStrategy(s *clusterv1.MachineDeploym
 	return m
 }
 
+// WithNamingStrategy sets the Strategy for the MachineDeploymentClassBuilder.
```
Comment (sbueringer) [inline internal/test/builder/builders.go:489]: Same for MP (pending merge/rebase) (looks like we don't have anything for ControlPlaneClass)

File: internal/controllers/topology/names/namegenerator.go [hunk 6]
```diff
@@ -0,0 +1,123 @@
+/*
```
Comment (sbueringer) [inline internal/controllers/topology/names/namegenerator.go:1]: Can we move this file into the cluster package?

File: internal/webhooks/clusterclass.go [hunk 7]
```diff
@@ -410,6 +414,41 @@ func validateMachineHealthCheckClasses(clusterClass *clusterv1.ClusterClass) fie
...
+		allErrs = append(allErrs, validateNamingStategy(md.NamingStrategy, field.NewPath("spec", "workers", "machineDeployments").Index(i).Child("namingStrategy"))...)
+	}
+	for i, mp := range clusterClass.Spec.Workers.MachinePools {
+		allErrs = append(allErrs, validateNamingStategy(mp.NamingStrategy, field.NewPath("spec", "workers", "machinePools").Index(i).Child("namingStrategy"))...)
+	}
+
+	return allErrs
+}
+
+func validateNamingStategy(namingStrategy *clusterv1.NamingStrategy, fldPath *field.Path) field.ErrorList {
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:431]:
```suggestion
func validateNamingStrategy(namingStrategy *clusterv1.NamingStrategy, fldPath *field.Path) field.ErrorList {
```

File: internal/webhooks/clusterclass.go [hunk 8]
```diff
@@ -410,6 +414,41 @@ func validateMachineHealthCheckClasses(clusterClass *clusterv1.ClusterClass) fie
...
+	if namingStrategy == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+
+	if namingStrategy.Name != nil {
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:438]: nit: maybe flip this check and add it to the if in l.432

File: internal/controllers/topology/names/namegenerator.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, g.data); err != nil {
+		return "", errors.Wrapf(err, "rendering template: %q", tpl.Name())
+	}
+
+	name := buf.String()
+
+	// If the name exceeds the maxNameLength: trim to maxGeneratedNameLength and add
+	// a random suffix.
+	if len(name) > maxNameLength {
```
Comment (sbueringer) [inline internal/controllers/topology/names/namegenerator.go:118]: We should probably document this behavior & the available variables via godoc on the API fields

File: api/v1beta1/clusterclass_types.go [hunk 10]
```diff
@@ -106,6 +106,10 @@ type ControlPlaneClass struct {
 	// +optional
 	MachineHealthCheck *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`
```
Comment (sbueringer) [inline api/v1beta1/clusterclass_types.go:108]: Let's set all 3 namingStrategies in clusterclass-quick-start-runtimesdk.yaml for CAPD e2e tests to a non-default template.

GIves us some happy-path e2e test coverage for free

File: api/v1beta1/clusterclass_types.go [hunk 11]
```diff
@@ -162,6 +166,10 @@ type MachineDeploymentClass struct {
 	// +optional
 	FailureDomain *string `json:"failureDomain,omitempty"`
 
+	// NamingStrategy allows to change the naming pattern used when creating the MachineDeployment.
+	// If not defined, it will fallback to `{{ .cluster.name }}-{{ .machineDeployment.topologyName }}-{{ .randomSuffix }}`
```
Comment (sbueringer) [inline api/v1beta1/clusterclass_types.go:170]:
```suggestion
	// If not defined, it will fallback to `{{ .cluster.name }}-{{ .machineDeployment.topologyName }}-{{ .random }}`
```
I think

File: api/v1beta1/clusterclass_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (killianmuldoon) [reply api/v1beta1/clusterclass_types.go:653]: Option 2 seems the cleanest to me as well - thanks for working on the examples!

[the remark this answers is not in the document]
Comment (fabriziopandini) [reply api/v1beta1/clusterclass_types.go:653]: I prefer 2 as well, but what about 

```
spec:
  ...
  controlPlane:
    namingStrategy:
      - field: metadata.name
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      - field:  spec.machineTemplate.infrastructureRef.Name
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"`
```

so we are a little bit more explicit about where the generated names will pop up in the controlPlane

[the remark this answers is not in the document]
Comment (sbueringer) [reply api/v1beta1/clusterclass_types.go:653]: This looks more like a generic engine to patch arbitrary fields to me to be honest

[the remark this answers is not in the document]
Comment (chrischdi) [reply api/v1beta1/clusterclass_types.go:653]: From discussions: there's a 4th option which would be more consistent compared to the current structure of the ClusterClass object:

```
spec:
  ...
  controlPlane:
    namingStrategy:
      template: "{{ .cluster.name }}-{{ .randomSuffix }}"
      machineInfrastructureTemplate:
        template: "{{ .cluster.name }}-{{ .randomSuffix }}"
```

[the remark this answers is not in the document]
Comment (sbueringer) [reply api/v1beta1/clusterclass_types.go:653]: Sounds good to me

[the remark this answers is not in the document]
Comment (killianmuldoon) [reply api/v1beta1/clusterclass_types.go:653]: +1 from me

File: api/v1beta1/clusterclass_types.go (hunk 10 above)
Comment (sbueringer) [reply api/v1beta1/clusterclass_types.go:108]: I'm a bit on the fence. On one side it's easy to do on the other side it's already covered by unit tests and we usually try to avoid validating too much details in the e2e tests.

I think I wouldn't validate it. Basically just happy path in the sense of the test still works/ cluster is coming still up

File: internal/webhooks/clusterclass.go [hunk 12]
```diff
@@ -157,6 +158,9 @@ func (webhook *ClusterClass) validate(ctx context.Context, oldClusterClass, newC
 	// Ensure MachineHealthChecks are valid.
 	allErrs = append(allErrs, validateMachineHealthCheckClasses(newClusterClass)...)
 
+	// Ensure NamePrefixes are valid templates.
+	allErrs = append(allErrs, validateNamingStrategies(newClusterClass)...)
```
Comment (chrischdi) [inline internal/webhooks/clusterclass.go:162]: TODO: "Generate" a name to check if valid.

File: internal/controllers/topology/cluster/namegenerator.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
+/*
```
Comment (chrischdi) [inline internal/controllers/topology/cluster/namegenerator.go:1]: Move to `internal/topology/<name>`

File: api/v1beta1/clusterclass_types.go [hunk 14]
```diff
@@ -106,6 +106,10 @@ type ControlPlaneClass struct {
 	// +optional
 	MachineHealthCheck *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`
 
+	// NamingStrategy allows to change the naming pattern used when creating the control plane provider object.
+	// If not defined, it will fallback to `{{ .cluster.name }}-{{ .random }}`.
+	NamingStrategy *ControlPlaneClassNamingStrategy `json:"namingStrategy,omitempty"`
```
Comment (sbueringer) [inline api/v1beta1/clusterclass_types.go:111]: Let's also flag all these fields with //+optional because they are optional (based on pointer & omitempty & semantics)

File: internal/controllers/topology/cluster/desired_state.go [hunk 15]
```diff
@@ -236,11 +235,16 @@ func (r *Reconciler) computeControlPlane(ctx context.Context, s *scope.Scope, in
...
+	if s.Blueprint.ClusterClass.Spec.ControlPlane.NamingStrategy != nil && s.Blueprint.ClusterClass.Spec.ControlPlane.NamingStrategy.Template != nil {
+		nameTemplate = *s.Blueprint.ClusterClass.Spec.ControlPlane.NamingStrategy.Template
+	}
+
 	controlPlane, err := templateToObject(templateToInput{
 		template:              template,
 		templateClonedFromRef: templateClonedFromRef,
 		cluster:               cluster,
-		namePrefix:            fmt.Sprintf("%s-", cluster.Name),
+		nameGenerator:         names.ControlPlaneNameGenerator(nameTemplate, s.Current.Cluster.Name),
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state.go:247]:
```suggestion
		nameGenerator:         names.ControlPlaneNameGenerator(nameTemplate, cluster.Name),
```
nit: I guess doesn't really matter but as we build that "shortcut" above

File: internal/controllers/topology/cluster/desired_state.go [hunk 16]
```diff
@@ -696,13 +707,23 @@ func computeMachineDeployment(ctx context.Context, s *scope.Scope, machineDeploy
...
 	}
 
+	nameTemplate := "{{ .cluster.name }}-{{ .machineDeployment.topologyName }}-{{ .random }}"
+	if machineDeploymentClass.NamingStrategy != nil && machineDeploymentClass.NamingStrategy.Template != nil {
+		nameTemplate = *machineDeploymentClass.NamingStrategy.Template
+	}
+
+	name, err := names.MachineDeploymentNameGenerator(nameTemplate, s.Current.Cluster.Name, machineDeploymentTopology.Name).GenerateName()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to generate name for machine deployment")
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/desired_state.go:717]:
```suggestion
		return nil, errors.Wrap(err, "failed to generate name for MachineDeployment")
```
I know it's a bit arbitrary how we name things in errors around here but let's try to use the actual names if we talk about core CAPI CRDs (same for MP below)


Let's also make sure we have consistent error wrapping for all generate cases (~ we should add errors.Wrap to  cp, bootstraptemplate, ...)

File: internal/test/builder/builders.go [hunk 17]
```diff
@@ -316,6 +316,7 @@ type ClusterClassBuilder struct {
 	controlPlaneNodeDeletionTimeout           *metav1.Duration
 	machineDeploymentClasses                  []clusterv1.MachineDeploymentClass
 	machinePoolClasses                        []clusterv1.MachinePoolClass
+	namingStrategy                            *clusterv1.ControlPlaneClassNamingStrategy
```
Comment (sbueringer) [inline internal/test/builder/builders.go:319]:
```suggestion
	controlPlaneNamingStrategy                            *clusterv1.ControlPlaneClassNamingStrategy
```
\+ move it up to the others

File: internal/test/builder/builders.go [hunk 18]
```diff
@@ -598,6 +615,9 @@ func (m *MachineDeploymentClassBuilder) Build() *clusterv1.MachineDeploymentClas
 	if m.strategy != nil {
 		obj.Strategy = m.strategy
 	}
+	if m.namingStrategy != nil {
```
Comment (sbueringer) [inline internal/test/builder/builders.go:618]: nit: up to you. Technically we can simplify a lot of this by getting rid of all the nil checking. Setting nil on a field that was never set before is a no-op. 

If you want to do it consistently across this file let's do it in a separate PR. But also fine to just keep it as is for me, just noticed it

File: internal/topology/names/names.go [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+import (
+	"bytes"
+	"fmt"
+	"text/template"
+
+	"github.com/pkg/errors"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+)
+
+const (
```
Comment (sbueringer) [inline internal/topology/names/names.go:29]: Let's just also add godoc here that these constants are taken from names.SimpleNameGenerator

File: internal/topology/names/names.go [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+func (g *templateGenerator) GenerateName() (string, error) {
+	tpl, err := template.New("template name generator").Option("missingkey=error").Parse(g.template)
+	if err != nil {
+		return "", errors.Wrapf(err, "parsing template %q", g.template)
+	}
+
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, g.data); err != nil {
+		return "", errors.Wrapf(err, "rendering template: %q", tpl.Name())
```
Comment (sbueringer) [inline internal/topology/names/names.go:115]: nit: will this read: "rendering template: template name generator: \<error\>". If yes, I would drop tpl.Name()

File: internal/topology/names/names.go [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	name := buf.String()
+
+	// If the name exceeds the maxNameLength: trim to maxGeneratedNameLength and add
+	// a random suffix.
+	if len(name) > maxNameLength {
```
Comment (sbueringer) [inline internal/topology/names/names.go:122]:
```suggestion
	if len(name) > maxGeneratedNameLength {
```

I wonder if we should use this for consistency. Just seems strange to use a different constant here compared to above

Comment (sbueringer) [reply internal/topology/names/names.go:122]: Otherwise we should probably use the same in SimpleNameGenerator but that would diverge from the upstream SimpleNameGenerator

Comment (sbueringer) [reply internal/topology/names/names.go:122]: If we change this, we should also update the godoc on the API

File: internal/webhooks/clusterclass.go [hunk 22]
```diff
@@ -410,6 +414,51 @@ func validateMachineHealthCheckClasses(clusterClass *clusterv1.ClusterClass) fie
...
+					fmt.Sprintf("invalid template: %v", err),
+				))
+		}
+	}
+
+	for i, md := range clusterClass.Spec.Workers.MachineDeployments {
+		if md.NamingStrategy == nil || md.NamingStrategy.Template == nil {
+			continue
+		}
+		if _, err := names.MachineDeploymentNameGenerator(*md.NamingStrategy.Template, "cluster", "topology").GenerateName(); err != nil {
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:435]:
```suggestion
		if _, err := names.MachineDeploymentNameGenerator(*md.NamingStrategy.Template, "cluster", "mdTopology").GenerateName(); err != nil {
```

maybe helps to avoid confusion just in case this somehow ends up in the reported error to users (same for MP)

File: internal/webhooks/clusterclass_test.go [hunk 23]
```diff
@@ -1165,6 +1165,50 @@ func TestClusterClassValidation(t *testing.T) {
...
+							builder.InfrastructureMachineTemplate(metav1.NamespaceDefault, "infra1").Build()).
+						WithBootstrapTemplate(
+							builder.BootstrapTemplate(metav1.NamespaceDefault, "bootstrap1").Build()).
+						WithNamingStrategy(&clusterv1.MachineDeploymentClassNamingStrategy{Template: pointer.String("template-{{ .cluster.name }}")}).
+						Build()).
+				Build(),
+			expectErr: false,
+		},
+		{
+			name: "should return error for invalid md namingStrategy.name",
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:1191]: maybe let's add one error case for CP & MP just so we're hitting these code paths

File: internal/webhooks/clusterclass_test.go [hunk 24]
```diff
@@ -1165,6 +1165,50 @@ func TestClusterClassValidation(t *testing.T) {
 				Build(),
 			expectErr: true,
 		},
+		{
+			name: "should not return error for valid md namingStrategy.name",
```
Comment (sbueringer) [inline internal/webhooks/clusterclass_test.go:1169]: let's add valid CP/MP strategies as well in this case

**Review by sbueringer — COMMENTED** [review]
just nits. nice work!

File: api/v1beta1/clusterclass_types.go [hunk 25]
```diff
@@ -106,6 +106,10 @@ type ControlPlaneClass struct {
 	// +optional
 	MachineHealthCheck *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`
 
+	// NamingStrategy allows to change the naming pattern used when creating the control plane provider object.
+	// If not defined, it will fallback to `{{ .cluster.name }}-{{ .random }}`.
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply api/v1beta1/clusterclass_types.go:110]: Yeah thats more correct 👍  the template field falls back 👍

File: internal/topology/names/names.go (hunk 21 above)
Comment (chrischdi) [reply internal/topology/names/names.go:122]: It is like this on purpose!

SimpleNameGenerator **always** adds a random suffix of length 5. So the input base string is not allowed to be longer than `maxGeneratedNameLength` (58), which results in a name of maximum `maxNameLength` (63).

The templateGenerator **only** trims and adds a random suffix of length 5, if the generated name exceeds the `maxNameLength` (63).

Comment (chrischdi) [reply internal/topology/names/names.go:122]: Note: the `{{ .random }}` could be anywhere in the template.

Comment (sbueringer) [reply internal/topology/names/names.go:122]: Yup. I assumed it was on purpose, but it's still an inconsistency between the generators :)

File: internal/webhooks/clusterclass.go [hunk 26]
```diff
@@ -157,6 +158,9 @@ func (webhook *ClusterClass) validate(ctx context.Context, oldClusterClass, newC
 	// Ensure MachineHealthChecks are valid.
 	allErrs = append(allErrs, validateMachineHealthCheckClasses(newClusterClass)...)
 
+	// Ensure NamePrefixes are valid templates.
```
Comment (sbueringer) [inline internal/webhooks/clusterclass.go:161]:
```suggestion
	// Ensure NamingStrategies are valid.
```
maybe?

Not that this adds additional information over the func name, but we have the same above 🗡️ 
namePrefixes seems off though

**Review by sbueringer — COMMENTED** [review]
Last round of nits, just godoc

**Comment by sbueringer** [comment]
lgtm pending verify

**Comment by sbueringer** [comment]
Thank you very much!

/lgtm

**Comment by sbueringer** [comment]
Quote from the issue:
> We would like to propose to cherry-pick this back into v1.5 (based on "Enhancements or additions to experimental Cluster API features"

/cherry-pick release-1.5

File: api/v1beta1/clusterclass_types.go [hunk 27]
```diff
@@ -106,6 +106,10 @@ type ControlPlaneClass struct {
 	// +optional
 	MachineHealthCheck *MachineHealthCheckClass `json:"machineHealthCheck,omitempty"`
 
+	// NamingStrategy allows to change the naming pattern used when creating the control plane provider object.
+	// +optional
```
Comment (killianmuldoon) [inline api/v1beta1/clusterclass_types.go:110]:
```suggestion
	// NamingStrategy allows changing the naming pattern used when creating the control plane provider object.
	// +optional
```

File: api/v1beta1/clusterclass_types.go [hunk 28]
```diff
@@ -162,6 +179,10 @@ type MachineDeploymentClass struct {
 	// +optional
 	FailureDomain *string `json:"failureDomain,omitempty"`
 
+	// NamingStrategy allows to change the naming pattern used when creating the MachineDeployment.
```
Comment (killianmuldoon) [inline api/v1beta1/clusterclass_types.go:182]:
```suggestion
	// NamingStrategy allows changing the naming pattern used when creating the MachineDeployment.
```

File: api/v1beta1/clusterclass_types.go [hunk 29]
```diff
@@ -267,6 +302,10 @@ type MachinePoolClass struct {
 	// +optional
 	FailureDomains []string `json:"failureDomains,omitempty"`
 
+	// NamingStrategy allows to change the naming pattern used when creating the MachinePool.
```
Comment (killianmuldoon) [inline api/v1beta1/clusterclass_types.go:305]:
```suggestion
	// NamingStrategy allows changing the naming pattern used when creating the MachinePool.
```

File: internal/controllers/topology/cluster/desired_state.go [hunk 30]
```diff
@@ -25,7 +25,6 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (killianmuldoon) [inline internal/controllers/topology/cluster/desired_state.go:25]: Fine as a follow-up, but we should document how to do this in the ClusterClass docs.

File: internal/controllers/topology/cluster/desired_state.go [hunk 31]
```diff
@@ -1230,7 +1261,11 @@ func templateToObject(in templateToInput) (*unstructured.Unstructured, error) {
 	// Ensure the generated objects have a meaningful name.
 	// NOTE: In case there is already a ref to this object in the Cluster, re-use the same name
 	// in order to simplify compare at later stages of the reconcile process.
```
Comment (killianmuldoon) [inline internal/controllers/topology/cluster/desired_state.go:1263]:
```suggestion
	// in order to simplify comparison at later stages of the reconcile process.
```

File: internal/controllers/topology/cluster/desired_state.go [hunk 32]
```diff
@@ -1290,12 +1325,16 @@ func templateToTemplate(in templateToInput) *unstructured.Unstructured {
 	// Ensure the generated template gets a meaningful name.
 	// NOTE: In case there is already an object ref to this template, it is required to re-use the same name
 	// in order to simplify compare at later stages of the reconcile process.
```
Comment (killianmuldoon) [inline internal/controllers/topology/cluster/desired_state.go:1327]:
```suggestion
	// in order to simplify comparison at later stages of the reconcile process.
```

**Review by killianmuldoon — COMMENTED** [review]
Congrats on the very straightforward implementation - I thought it would be much more complex.

Just a couple of nits and a follow up from my side.

File: internal/controllers/topology/cluster/desired_state.go (hunk 30 above)
Comment (sbueringer) [reply internal/controllers/topology/cluster/desired_state.go:25]: Which part of the docs do you think would this fit into?

I think for other fields we usually didn't write docs apart from godocs / descriptions for the CRD

Comment (chrischdi) [reply internal/controllers/topology/cluster/desired_state.go:25]: I'd do this in a follow-up PR :-)

Comment (sbueringer) [reply internal/controllers/topology/cluster/desired_state.go:25]: Yup let's just discuss first what we want to document where before we do it ;)

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve
